app/models/generic_event.rb

```diff
@@ -143,14 +143,43 @@ def self.details_attributes(*attributes)
     end
   end
 
+  # NB: Majority of events will use this serializer rather than the anonymous class.
+  def self.serializer
+    GenericEventSerializer
+  end
+
   # Relationship attributes live against the details but are expected in the json:api relationship section
   # so are defined separately
   def self.relationship_attributes(attributes)
     define_singleton_method(:relationship_attributes) do
       instance_variable_get('@relationship_attributes')
     end
+
     instance_variable_set('@relationship_attributes', attributes)
 
+    define_singleton_method(:serializer) do
+      @serializer ||=
+        begin
+          serializer = Class.new(GenericEventSerializer)
+
+          relationship_attributes.each do |attribute_key, attribute_type|
+            # NB: Default to V2 relationship serializer and fallback to unversioned serializer. Never support v1
+            relation_serializer = "V2::#{attribute_type.to_s.singularize.camelize}Serializer"
+            relation_serializer = if const_defined?(relation_serializer)
+                                    relation_serializer
+                                  else
+                                    relation_serializer.gsub('V2::', '')
+                                  end
+
+            named_relationship_key = attribute_key.to_s.sub('_id', '')
+            serializer.set_type :events
```

Comment:
This is necessary as the type is accessed via instance variable in the parent and needs to be set in the child.

Reply:
Really not convinced you need this line - in `GenericEventSerializer` there is `set_type :events` method call, Maybe I'm wrong - could you try without and see if it still works? I couldn't spot any coverage that proves this is needed (unless there's some existing stuff) so might be worth adding an explicit check on `type`.

Reply:
I'll pull out the exact line of code in the serializer core that explains this. I didn't add it for fun unfortunately. There might be another way of achieving it (overriding a method in the parent and using the method lookup tree, for example)

Reply:
It seems strange that the inherited type is working fine for this class - https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/master/app/serializers/allocations_serializer.rb - and tested explicitly here https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/master/spec/requests/api/allocations_controller_index_spec.rb#L30

Quite the mystery 👻 🐕 🚗

Reply:
Here's a simpler spec that shows the same thing: https://github.com/ministryofjustice/hmpps-book-secure-move-api/blob/master/spec/serializers/allocations_serializer_spec.rb#L17

I wonder if the behaviour you're seeing is just down to the dynamic class definitions rather than static approach for allocations. It's more academic curiosity at this point to be honest, happy for the line to be left in place and we can ponder this some other time.

Reply:
Discussed offline and solved the mystery - in the case of allocations serializer the `type` is defaulted to `allocations` entirely based on the class name, and the `type` is indeed not inherited from the ancestor. Move along people, nothing to see here 😁